include/KeyTable.h

```diff
@@ -0,0 +1,170 @@
+#pragma once
+
+/* a lot of this code was stolen from https://github.com/kernc/logkeys but then modified */
+
+#include <cstring>
+#include <string>
+#include <vector>
+
+#include <linux/input.h>
+
+namespace hemiola
+{
+    // forward declaration
+    struct KeyState;
+
+    /*!
+     * @brief class which contains the system key map
+     */
+    class KeyTable
+    {
+    public:
+        KeyTable() = default;
+        ~KeyTable() = default;
+
+        /*!
+         * @brief Update our keymaps based on the system keymapping
+         * @post keymaps are updated to coincide with the system keymap
+         */
+        void determineSystemKeymap();
+
+        /*!
+         * @brief check if code is a char
+         * @param code the scan code from key press
+         * @return true if code is a char
+         */
+        bool isCharKey ( unsigned int code ) const;
+
+        /*!
+         * @brief check if code is a function key
+         * @param code the scan code from key press
+         * @return true if code is a function key
+         */
+        bool isFuncKey ( unsigned int code ) const;
+
+        /*!
+         * @brief check if code is used
+         * @param code the scan code from key press
+         * @return true if code is used
+         */
+        inline bool isUsedKey ( unsigned int code ) const;
+
+        /*!
+         * @brief convert a scan code to a character
+         * @param code the scan code from key press
+         * @param keyState the current state of the keypress
+         * @return character representation of key press
+         */
+        wchar_t handleScanCode ( unsigned int code, const KeyState& keyState ) const;
+
+        /*
+         * @brief check if the scan code is a valid scan code
+         * @param code the scan code from key press
+         * @return true if the code is valid
+         */
+        bool isCodeValid ( unsigned int code ) const { return code < m_CharOrFunc.size(); }
+
+    private:
+        /*!
+         * @brief get the lower case character corresponding to a scan code
+         * @param code the scan code from key press
```

Comment:
Nit: may want to extend the param or return to say that the it's a pass-through if a conversion can't be done.

Reply:
For here there is no pass through. I will put a comment elsewhere that discusses that.